File: 1146-greatest-common-divisor-of-strings/1146-greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStringsSlow(self, str1: str, str2: str) -> str:
        s = str1 if len(str1) < len(str2) else str2
        t = ""
        ret = ""
        for c in s:
            t +=c
            if str1.replace(t,"") == "" and str2.replace(t,"") == "":
                ret = t
        return ret

    def gcdOfStrings(self, str1: str, str2: str) -> str:

        if str1 + str2 != str2 + str1:
            return ""


        def gcd(a, b):
            while b:
                a,b = b,a%b
            return a
        
        gcd_length = gcd(len(str1), len(str2))

        return str1[:gcd_length]
        

File: 1146-greatest-common-divisor-of-strings/1146-greatest-common-divisor-of-strings.cpp
class Solution {
public:
    string gcdOfStringsMem(string str1, string str2) {
        if (str1 + str2 != str2 + str1) {
            return "";
        }
        function<int(int, int)> gcd = [](int a, int b) {
            while (b != 0) {
                int temp = b;
                b = a % b;
                a = temp;
            }
            return a;
        };
        int gcdLength = gcd(str1.size(), str2.size());
        return str1.substr(0, gcdLength);        
    }

    string gcdOfStrings(string str1, string str2) {
        return str1+str2==str2+str1?str2.substr(0,gcd(str1.length(),str2.length())):"";
    }
};
